---
output: pdf_document
---
# Conducting Linear Regression Based GWAS
### Introduction
The goal of this exampel is to demonstrate how you can use functions within the 
Staphopia API to conduct a GWAS. For this demonstration we will be using an 
already published 

### Let's Begin
#### Import Packages
```{r}
library(staphopia)
library(ggplot2)
```

#### Minimum Allele Frequency
We can set a constant for minimum allele frequency (MAF) at which we'll filter
out the rare SNPs.
```{r}
MAF <- 0.05
```

#### Acquiring the VISA tag
For ease of access Alam et al. 2014 has been stored under the Tag 
'visa-gwas-2014'. The first thing we'll need to do is get information 
(id, comment, etc...) about the tag using the `get_tag_by_name` by name 
function.
```{r}
tag <- get_tag_by_name('visa-gwas-2014')
tag
```

#### Acquiring all samples associated with tag 'visa-gwas-2014'
Using the id of the tag (visa-gwas-2014) which we just acquired, we can 
easily pull down all the samples assocaited with tag with the 
`get_sampels_by_tag` function.
```{r}
samples <- get_samples_by_tag(tag$id)
head(samples)
```

#### Now the Phenotype
These data tested the phenotype of 74 *S. aureus* samples against Vancomycin. 
In this demonstration we will use the minimum inhibitatory concentrations 
(MIC ug/ml) as determined by the Etest.

First we will use `get_resistance` and specify the *antibiotic* as vancomycin 
and the *test* as Etest.
```{r}
vancomycin <- get_resistance(antibiotic="vancomycin", test="etest")
vancomycin
```

Now using the resistance id we just acquired, we can pull the resistance 
phenotype for each of samples but only pull down those associated with 
Vancomycin Etest. This is done using `get_resistance_by_sample` and setting 
the *resistance_id* parameter.
```{r}
phenotype <- get_resistance_by_samples(samples$sample_id,
                                       resistance_id=vancomycin$id)
head(phenotype)
```


#### Merge phenotype and samples, then sort by sample_tag
Nothing much going on here. In this step we are simply merging the phenotype 
data frame into the samples data frame. Then we sort the samples data frame by
*sample_tag*.
```{r}
samples <- merge(samples, phenotype)
samples <- samples[order(samples$sample_tag),]
head(samples)
```

#### Acquiring all SNPs in each sample
We're conducting a GWAS, so we probably need some SNPs. There are two ways to 
get the SNPs associated with a sample. You can get them by sample with 
`get_snps(sample_id)` or you can get them by all samples at once using 
`get_snps_by_samples(sample_ids)`. In our case since we have 74 samples, we'll 
be using `get_snps_by_samples`.
```{r}
snps <- get_snps_by_samples(samples$sample_id)
head(snps)
```

#### Annotating *snp_id*
As you might notice we have a list of samples and snps, but they're only ids. 
There's not much biological information behind these ids. So let's use the 
`get_snps_in_bulk` to pull down all of the SNPs at once. In order to do so we 
must pass *snps$snp_id*. It will take care of only pulling the unique snp_ids.
```{r}
snp_info <- get_snps_in_bulk(snps$snp_id)
head(snp_info[,1:8])
```

#### Present/Absent SNP matrix
For our GWAS we'll need to build a simple matrix. In this matrix each column 
will represent a sample and each row a SNP position. Now for each SNP we need 
to determine if its present (1) or absent (0) in each sample. No worries! 
There's a function for that! Using `create_snp_matrix`, we can pass all the 
info we've acquired *snps*, *samples* and *snp_info*. Leaving us with something 
like the following.
```{r}
snp_matrix <- create_snp_matrix(snps, samples, snp_info)
head(snp_matrix[,1:6])
```

#### Conducting a linear regression based GWAS
Now the moment we are all here for, GWAS! We'll be using R's bnuilt in `lm` 
(linear regression) function. It produces the same output as PLINK, just a bit 
slower. In order to do so, we can use the `run_gwas` function and pass our 
*snp_matrix* and the phenotype (*sample$mic*) to be tested.
```{r}
gwas <- run_gwas(snp_matrix, samples$mic)
head(gwas)
```

#### Remove SNPs < MAF
Rare SNPs may produce spurious results, so lets remove those SNPs that are at a 
frequency less than our minimum allele frequency (*MAF*) which we set earlier.
```{r}
MAF

# Pre-filter length
nrow(gwas)
gwas <- gwas[gwas$freq >= MAF,]

# Post-filter length
nrow(gwas)
head(gwas)
```

#### Visualize the results
We can use `manhattan_plot` and `qq_plot` to visualize the results of our GWAS.
By default each of the plots will use Bonferroni correction to determine 
significant SNPs.

##### Manhattan Plot
```{r}
manhattan_plot(gwas)
```

# QQ Plot
```{r}
qq_plot(gwas[with(gwas, order(logp)), ])
```

#### The End

